### ISSUE TITLE
POSIX::PIPE_BUF is not properly defined

### ISSUE THREAD

**Issue by sidney** [StrawberryPerl/Perl-Dist-Strawberry#40]
The limits.h header in mingw32 sets PIPE_BUF to 512, which works fine in programs I have tried. Strawberry Perl should be using that value. It is currently left undefined as shown below:

>perl -v
This is perl 5, version 32, subversion 1 (v5.32.1) built for MSWin32-x64-multi-thread

>perl -MPOSIX=PIPE_BUF -e "print PIPE_BUF;"
Your vendor has not defined POSIX macro PIPE_BUF, used at -e line 1

**Comment by sisyphus** [comment]
<code>PIPE_BUF</code> (and some other symbols) is defined in limits.h only if <code>\_POSIX\_</code> is defined.
Currently, in the perl build <code>\_POSIX\_</code> is not defined when limits.h is processed at the beginning of ext/POSIX/POSIX.xs.
limits.h is pulled in by perl.h, so <code>\_POSIX\_</code> needs to be defined before perl.h is included.
Here's a patch to POSIX.xs that does this if and only if perl is being built by a mingw toolchain.
```
--- POSIX.xs_orig       2022-02-24 13:35:27 +1100
+++ POSIX.xs    2022-06-15 14:26:17 +1000
@@ -1,6 +1,10 @@
 #define PERL_EXT_POSIX
 #define PERL_EXT

+#if defined(_WIN32) && defined(__GNUC__) /* mingw compiler */
+#define _POSIX_
+#endif
+
 #define PERL_NO_GET_CONTEXT

 #include "EXTERN.h"
```
That works well for me.
However, I'm not sure that the Strawberry project wants to get involved in using patched perl source.

The best thing would be to have the official perl source fixed to address this issue - and the best chance of getting that done is to open an issue at https://github.com/Perl/perl5/issues or a pull request at https://github.com/Perl/perl5/pulls .

Cheers,
Rob

**Comment by sidney** [comment]
Perfect, I'll submit the PR/issue there and close this issue.

**Comment by sisyphus** [comment]
@sidney, I think that defining \_POSIX\_ in POSIX.xs limits the scope of the definition to the POSIX module, so it should be safe.
I've tried that patch in the perl-5.36.0 source. Perl-5.36.0 then builds just fine, and the perl-5.36.0 test suite still passes, too.

I, too, had wondered about adding <code>-D\_POSIX\_</code> to buildoptextra.
That would, of course, make the \_POSIX\_ definition visible just about everywhere.
Anywat, now that you've tried it out, and found it to be wanting, I no longer need to wonder about it ;-)

Also note that it's not just PIPE_BUF that we're making available. There's a bunch of other POSIX constants in limits.h that are also being made visible.

Another approach would be to just insert those limits.h definitions into POSIX.xs:
```
#if defined(__MINGW64_VERSION_MAJOR) /* 32-bit or 64-bit mingw-w64 compiler */
#define _POSIX_ARG_MAX 4096
#define _POSIX_CHILD_MAX 6
#define _POSIX_LINK_MAX 8
#define _POSIX_MAX_CANON 255
#define _POSIX_MAX_INPUT 255
#define _POSIX_NAME_MAX 14
#define _POSIX_NGROUPS_MAX 0
#define _POSIX_OPEN_MAX 16
#define _POSIX_PATH_MAX 255
#define _POSIX_PIPE_BUF 512
#define _POSIX_SSIZE_MAX 32767
#define _POSIX_STREAM_MAX 8
#define _POSIX_TZNAME_MAX 3
#define ARG_MAX 14500
#define LINK_MAX 1024
#define MAX_CANON _POSIX_MAX_CANON
#define MAX_INPUT _POSIX_MAX_INPUT
#define NAME_MAX 255
#define NGROUPS_MAX 16
#define OPEN_MAX 32
#undef PATH_MAX
#define PATH_MAX 512
#define PIPE_BUF _POSIX_PIPE_BUF
#define STREAM_MAX 20
#define TZNAME_MAX 10
#endif
```
The problem there being that, if a symbol changes values or is removed, or if new symbols are added, then we need to correct POSIX.xs. (And we might not even be aware that limits.h has made such changes.)
I think it's preferable that they be read in directly from limits.h.

Cheers,
Rob

**Comment by sidney** [comment]
I submitted a pr https://github.com/Perl/perl5/pull/19866
I'll close this if they respond positively.

**Comment by sisyphus** [comment]
Cool - let's see what happens.

Of course, if that patch is applied to the perl source, it's still not going to appear in a stable release until 5.36.1 or 5.38.0.
In the meantime, Strawberry will hopefully be releasing 5.34.0, 5.34.1 and 5.36.0.

I don't rule out the possibility that Strawberry will opt to include that patch in those releases.
I don't know what their policy on such things is (if there even is one).

So I think you could leave this issue open - at least until someone states authoritatively that this patch either will be or won't be included in those Strawberry builds.

Cheers,
Rob

**Comment by sidney** [comment]
For now my code looks something like $RUNNING_ON_WINDOWS?512:POSIX_PIPE and that's good enough. At least this is now documented as an open issue.

**Comment by sidney** [comment]
The PR has been merged into blead on Perl/perl5 so it's there for the future.
As you suggested, I'll leave this as an open issue in case somebody on the Strawberry team wants to incorporate it into an earlier future release.

**Comment by shawnlaffan** [comment]
This is in perl 5.38.0 and thus is part of Strawberry Perl 5.38.0.1.  It's not in perl 5.36.1, and was not patched across to the SP 5.36.1.1 release (sorry).  

https://github.com/Perl/perl5/commit/491d7f25f50e7e1c26ca3e9793ce5a6df78c4163

Is there a pressing need for it to be in a future SP 5.36.x release?  If not then we can close this issue.

**Comment by shawnlaffan** [comment]
There's been no response since my last post so I assume there is no pressing need for this to be in Strawberry Perl 5.36.  I'll close this issue now.
